salt/modules/nxos.py

```diff
@@ -1,49 +1,217 @@
+    - 'feature ssh' must be enabled for SSH proxy minions.
+    - 'feature nxapi' must be enabled for NX-API proxy minions.
+    - Settings in the proxy minion configuration file must match device settings.
+    - NX-OS device is reachable from the Salt Master.
+"""
+
 
 def __virtual__():
+    return __virtualname__
+
+
+def ping(**kwargs):
+    """
+    Ping the device on the other end of the connection.
+
+    .. code-block: bash
+
+        salt '*' nxos.cmd ping
+    """
     if salt.utils.platform.is_proxy():
-        return __virtualname__
-    return (
-        False,
-        "The nxos execution module failed to load: only available on proxy minions.",
-    )
+        return __proxy__["nxos.ping"]()
+    return __utils__["nxos.ping"](**kwargs)
+
+
+# -----------------------------------------------------------------------------
+# Device Get Functions
+# -----------------------------------------------------------------------------
+def check_password(username, password, encrypted=False, **kwargs):
+    """
+    Verify user password.
+
+    username
+        Username on which to perform password check
+
+    password
+        Password to check
+
+    encrypted
+        Whether or not the password is encrypted
+        Default: False
+
+    .. code-block: bash
 
+        salt '*' nxos.cmd check_password username=admin password=admin
+        salt '*' nxos.cmd check_password username=admin \\
+            password='$5$2fWwO2vK$s7.Hr3YltMNHuhywQQ3nfOd.gAPHgs3SOBYYdGT3E.A' \\
+            encrypted=True
+    """
+    hash_algorithms = {
+        "1": "md5",
+        "2a": "blowfish",
+        "5": "sha256",
+        "6": "sha512",
+    }
+    password_line = get_user(username, **kwargs)
+    if not password_line:
+        return None
+    if "!" in password_line:
+        return False
+    cur_hash = re.search(r"(\$[0-6](?:\$[^$ ]+)+)", password_line).group(0)
+    if encrypted is False:
+        hash_type, cur_salt, hashed_pass = re.search(
+            r"^\$([0-6])\$([^$]+)\$(.*)$", cur_hash
+        ).groups()
+        new_hash = gen_hash(
+            crypt_salt=cur_salt,
+            password=password,
+            algorithm=hash_algorithms[hash_type],
+            force=True,
+        )
+    else:
+        new_hash = password
+    if new_hash == cur_hash:
+        return True
+    return False
 
-def system_info():
+
+def check_role(username, role, **kwargs):
     """
-    Return system information for grains of the NX OS proxy minion
+    Verify role assignment for user.
 
     .. code-block:: bash
 
-        salt '*' nxos.system_info
+        salt '*' nxos.cmd check_role username=admin role=network-admin
     """
-    return cmd("system_info")
+    return role in get_roles(username, **kwargs)
 
 
 def cmd(command, *args, **kwargs):
     """
-    run commands from __proxy__
-    :mod:`salt.proxy.nxos<salt.proxy.nxos>`
+    NOTE: This function is preserved for backwards compatibilty.  This allows
```

Comment:
@mikewiebe: which is the preferred form? It seems that `nxos.function` is preferred, but most of the documentation for the commands says to use `nxos.cmd <function>`?

Reply:
@dhiltonp `nxos.function` is definitely preferred.  I am updating the docs now.